services/wallet/transfer/reactor.go

```diff
@@ -60,35 +85,216 @@ func (r *Reactor) newControlCommand(chainClient *chain.ClientWithFallback, accou
 	return ctl
 }
 
-// Start runs reactor loop in background.
-func (r *Reactor) start(chainClients map[uint64]*chain.ClientWithFallback, accounts []common.Address) error {
-	r.mu.Lock()
-	defer r.mu.Unlock()
+func (s *OnDemandFetchStrategy) start() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
 
-	if r.group != nil {
+	if s.group != nil {
 		return errAlreadyRunning
 	}
-	r.group = async.NewGroup(context.Background())
-	for _, chainClient := range chainClients {
-		ctl := r.newControlCommand(chainClient, accounts)
-		r.group.Add(ctl.Command())
+	s.group = async.NewGroup(context.Background())
+
+	for _, chainClient := range s.chainClients {
+		ctl := s.newControlCommand(chainClient, s.accounts)
+		s.group.Add(ctl.Command())
 	}
+
 	return nil
 }
 
 // Stop stops reactor loop and waits till it exits.
-func (r *Reactor) stop() {
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	if r.group == nil {
+func (s *OnDemandFetchStrategy) stop() {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.group == nil {
 		return
 	}
-	r.group.Stop()
-	r.group.Wait()
-	r.group = nil
+	s.group.Stop()
+	s.group.Wait()
+	s.group = nil
+}
+
+func (c *OnDemandFetchStrategy) kind() FetchStrategyType {
+	return OnDemandFetchStrategyType
+}
+
+func (s *OnDemandFetchStrategy) getTransfersByAddress(ctx context.Context, chainID uint64, address common.Address, toBlock *big.Int,
+	limit int64, fetchMore bool) ([]Transfer, error) {
+
+	log.Info("[WalletAPI:: GetTransfersByAddress] get transfers for an address", "address", address, "fetchMore", fetchMore,
+		"chainID", chainID, "toBlock", toBlock, "limit", limit)
+
+	rst, err := s.db.GetTransfersByAddress(chainID, address, toBlock, limit)
+	if err != nil {
+		log.Error("[WalletAPI:: GetTransfersByAddress] can't fetch transfers", "err", err)
+		return nil, err
+	}
+
+	transfersCount := int64(len(rst))
+	chainClient, err := getChainClientByID(s.chainClients, chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	if fetchMore && limit > transfersCount {
+
+		block, err := s.blockDAO.GetFirstKnownBlock(chainID, address)
+		if err != nil {
+			return nil, err
+		}
+
+		// if zero block was already checked there is nothing to find more
+		if block == nil || big.NewInt(0).Cmp(block) == 0 {
+			log.Info("[WalletAPI:: GetTransfersByAddress] ZERO block is found for", "address", address, "chaindID", chainID)
+			return rst, nil
+		}
+
+		from, err := findFirstRange(ctx, address, block, chainClient)
+		if err != nil {
+			if nonArchivalNodeError(err) {
+				if s.feed != nil {
+					s.feed.Send(walletevent.Event{
+						Type: EventNonArchivalNodeDetected,
+					})
+				}
+				from = big.NewInt(0).Sub(block, big.NewInt(100))
```

Comment:
This `100` represents a chunk/page size, right? Not sure if it will be reused in the new strategy, either way a const could be defined for clarity.

Reply:
Wondering where we check that this substraction won't generate invalid block numbers